### DISCUSSION TITLE
How to mock/handle refetch in jest?

### DISCUSSION THREAD

**Discussion by shapeshifta78** [TanStack/query#1347] [Q&A]
What is the proper way to handle the refetch of queries in jest? E.g. we define a query via usequery and trigger a refetch in an eventhandler with new parameters. How would I mock that in this case so the test in jest gets the new data?

**Comment by DannyDelott** [comment]
Not sure how rigid your requirements are for refetching in the event handler, but you could just invalidate the query in the handler then let your component-level `useQuery()` managing refetching for you.

**Reply by shapeshifta78** [reply]
That is what we do, but what would a test in jest look like?
Maybe I got you wrong?

**Reply by DannyDelott** [reply]
Ah, I thought you're describing actually calling the `refetch` function in an event handler, ie:

```tsx
const SomeComponent = () => {
  const { data, refetch } = useQuery('my-key', ...);

  // somewhere else, possibly in a useEffect
  Foo.on('my-event', () => refetch());
}
```

vs


```tsx
const SomeComponent = () => {
  const { data } = useQuery('my-key', ...);

  // somewhere else, possibly in a useEffect
  Foo.on('my-event', () => queryCache.invalidateQueries('my-key'));
}
```

**Reply by DannyDelott** [reply]
If you're using the second example, then you could just make assertions against the query cache, with the assumption that `useQuery` will work as expected and therefore need not be tested:

```ts
test('my test', () => {
  // ...do thing that triggers event handler

  // if this is true, then useQuery should Just Work 
  expect(queryCache.getQuery('my-key').isStale()).toEqual(true);
})
```
